fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,431 @@
+            )
+```
+
+#### `get_remaining_weight_in_epoch`
+
+```python
+def get_remaining_weight_in_epoch(store: Store, current_slot: Slot) -> Gwei:
+    # Returns the total weight of votes for this epoch from future committees after the current slot
+    first_slot_next_epoch = compute_start_slot_at_epoch(Epoch(compute_epoch_at_slot(current_slot) + 1))
+    return get_committee_weight_between_slots(store, Slot(current_slot + 1), Slot(first_slot_next_epoch - 1))
+```
+
+#### `get_leaf_block_roots`
+
+```python
+def get_leaf_block_roots(store: Store, block_root: Root) -> Set[Root]:
+    children = [
+        root for root in store.blocks.keys()
+        if store.blocks[root].parent_root == block_root
+    ]
+
+    if any(children):
+        # Get leaves of all children and add to the set.
+        leaves = set().union(*[get_leaf_block_roots(store, child) for child in children])
+        return leaves
+    else:
+        # This block is a leaf.
+        return set(block_root)
+
+```
+
+#### `get_ffg_support`
+
+```python
+def get_ffg_support(store: Store, block_root: Root) -> Gwei:
+    # Returns the total weight supporting the highest checkpoint in the block's chain
+
+    block = store.blocks[block_root]
+    assert get_current_store_epoch(store) == compute_epoch_at_slot(block.slot)
+
+    leave_roots = get_leaf_block_roots(store, block_root)
+    # current_epoch_attestations contains only attestations with source matching block.current_justified_checkpoint
+    attestations_in_leaves = set().union(
+        *[store.block_states[root].current_epoch_attestations for root in leave_roots]
+    )
+
+    current_epoch = get_current_store_epoch(store)
+    checkpoint_root = get_checkpoint_block(store, block_root, current_epoch)
+    support_for_checkpoint = {a for a in attestations_in_leaves if a.data.target.root == checkpoint_root}
+    checkpoint_state = store.block_states[checkpoint_root]
+    return get_attesting_balance(checkpoint_state, list(support_for_checkpoint))
+```
+
+#### `is_ffg_confirmed`
+
+```python
+def is_ffg_confirmed(
+    # Returns whether the branch will justify it's current epoch checkpoint at the end of this epoch
+    store: Store,
+    confirmation_byzantine_threshold: int,
+    confirmation_slashing_threshold: int,
+    block_root: Root,
+) -> bool:
+    current_slot = get_current_slot(store)
+    block = store.blocks[block_root]
+    assert get_current_store_epoch(store) == compute_epoch_at_slot(block.slot)
+
+    current_epoch = get_current_store_epoch(store)
+    checkpoint_root = get_checkpoint_block(store, block_root, current_epoch)
+    checkpoint_state = store.block_states[checkpoint_root]
+
+    remaining_ffg_weight = int(get_remaining_weight_in_epoch(store, current_slot))
+    total_active_balance = int(get_total_active_balance(checkpoint_state))
+    current_weight_in_epoch = total_active_balance - remaining_ffg_weight
+    assert current_weight_in_epoch >= 0
```

Comment:
Yep. You are right. This is a bug. `get_remaining_weight_in_epoch` should use the same state as `get_total_active_balance`

Reply:
if the justified state is ok to use, it would make the logic easier. otherwise, in implementations, fork choice context needs to be extended to track historical total active balance for each node

Reply:
Fixed by 1269ea56837b6fa3b2e0929b4287cd24fb71cda9
Unfortunately, the justified state is not safe to use.
Is the state of the latest checkpoint (see commit) expensive to use?

Reply:
It's alright to use it, just means that have to cache `get_total_active_balance` for non-finalized blocks. Not something that's currently being tracked, but also quite easy to start tracking.

Reply:
There was a bug in my previous commit, which I have (hopefully) fixed in b9919b6118af5c5961e3b31e0f93f3de88057175

`checkpoint_states` should contain the pulled-up state. Right?
See `store_target_checkpoint_state`

Reply:
Ah, right. It is now `store.checkpoint_states` instead of `store.block_states`. So, yeah, should be the pulled state now.

For the purpose of `get_total_active_balance` and `get_active_validator_indices`, `store.checkpoint_states[checkpoint]` should give the same result as `store_block_states[block_root]`, no? Like, I'm not understanding yet why we even need to fetch that checkpoint.

Reply:
>  I'm not understanding yet why we even need to fetch that checkpoint.

See [this comment](https://github.com/ethereum/consensus-specs/pull/3339/files#r1230360047)

Reply:
Thanks for pointing this out.
The code has now been changed to use the block's state to fetch balances.